Add initial files.

LukasNorbutas opened the pull request:
- Currently bin/entrypoint.sh does not serve any substantial purpose, as pre-commit can be installed in the Dockerfile as well. Any alternative suggestions, or should we keep it there as a showcase?
- Any suggestions for pre-commit hooks? Currently runs black, mypy, several hooks from pre-commit just to hint people where to look for more.

astanzys-vinted commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM continuumio/miniconda3
```
Some notes on the docker image:

* `entrypoint.sh` sets up pre commit hooks on local developer machine. It should be renamed accordingly and removed from the docker image.
* We should copy only those files that are needed to run the component. Docker images are deployment artifacts, so they should only include files needed to run the component. It should be enough to include 'src' and 'requirements.txt'.
* There's no need to make python files executable if you run them using `python`. It's only necessary if you want to run them like a shell script, i.e. `src/preprocessor/standard_scaler.py`
* Current `Dockerfile` is not cache friendly. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/ , `Levering build cache`. TLDR; When building the image you will have to redownload dependencies every time you change component code. The proper way is to copy requirements.txt, install dependencies, then copy source files.

Allardvm replied:
The usual way to handle the cache friendly Dockerfiles with Python code is to split the install into multiple stages. The first stage _only_ copies the `requirements.txt` and/or `environment.yml` to the image, then installs the requirements. The second stage copies the application source into the image and installs it—usually in development mode (`pip install -e .`).

Allardvm replied:
I would also not use `miniconda3` as base image, since it's large and doesn't give us an easy way to use a specific Python version. My vote would be to see if we can get a reliable setup with Google's distroless containers in combination with miniconda.

LukasNorbutas replied:
I addressed the caching issue. For now, we decided to keep the base image, since it is not too urgent. I will make a task for this and include it in the next iterations.

astanzys-vinted commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,66 @@
+    )
+    args = parser.parse_args()
+
+    # Run component logic here
```
Generally speaking, comments should explain the "why", not the "what". Here you could extract a function and remove the comment. Comments tend to become outdated unless extra care is taken. Code is always up to date, because otherwise your stuff breaks.

astanzys-vinted commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,66 @@
+
+# Main logic of the component
+if __name__ == "__main__":
+    logging.getLogger().setLevel(logging.INFO)
```
I think it would be nice to have a separate file to hold the entry point of the program. `main.py` or `entry.py` would work. It's just nice being able to find the entry point of a program fast.

Allardvm replied:
Agree. This'd be the typical use case for a `main.py` or `entrypoint.py`.

LukasNorbutas replied:
Done! Exported this part to entrypoint.py

astanzys-vinted commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,66 @@
+    input_data = read_csv_to_df(args.input_path)
+    input_data_preprocessed = standard_scaler(input_data, cols=target_cols)
+
+    if not args.output_path.startswith("gs://"):
```
Could you explain why this check is needed? If we have a google storage URI our program will crash anyways later on.

LukasNorbutas replied:
AFAIR, pandas to_csv method creates dirs when you write to cloud bucket (uses google-cloud-storage lib), but does not for local paths. I will keep this for now, will check this if this really holds and edit if necessary.

astanzys-vinted commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,66 @@
+    return data_copy
+
+
+def write_df_to_csv(input_df: pd.DataFrame, output_path: str, *args, **kwargs) -> None:
```
I would suggest removing this function. It provides no additional functionality over the wrapped `input_df.to_csv` method

LukasNorbutas replied:
Removed read/write wrapper functions. Now only entrypoint.py contains I/O functions, while the main script only contains the scaler.

astanzys-vinted commented on tests/conftest.py:
```diff
@@ -0,0 +1,40 @@
+            item.add_marker(skip_slow)
+
+
+@pytest.fixture
```
I'm not familiar with pytest, but these fixtures don't seem general enough to be included here. Could they be moved to the relevant test?

LukasNorbutas replied:
For now, deleted fixtures and moved constants either to test_scaler.py or replaced them within test functions. Will use fixtures when we include more tests.

astanzys-vinted commented on tests/test_scaler.py:
```diff
@@ -0,0 +1,38 @@
+import pytest
+from pytest import approx
+
+from google.cloud import storage
```
unused

astanzys-vinted commented on src/preprocessor/__init__.py:
```diff
@@ -0,0 +1 @@
+from .standard_scaler import read_csv_to_df, standard_scaler, write_df_to_csv
```
I'm still quite new to python, so maybe you could explain how this is helpful? My `__init__.py` files are always empty

Allardvm replied:
There's often no need for this. Complex packages use this to enable easy important of deeply nested functions and modules. The usage here would enable `from preprocessor import read_csv_to_df`.

Allardvm replied:
I'd recommend _not_ using these relative imports, since they are far more likely to lead to accidentally broken code (depending on the working directory, Python version, etc.). I.e. always import with `from preprocessor.standard_scaler import A, B, C`.

LukasNorbutas replied:
Done

astanzys-vinted commented on tests/test_scaler.py:
```diff
@@ -0,0 +1,38 @@
+    tests_root = os.path.abspath(os.path.dirname(__file__))
+    test_df = read_csv_to_df(os.path.join(tests_root, "test_data/test.csv"))
+    write_df_to_csv(test_df, os.path.join(tmpdir, "test_output.csv"))
+    assert os.path.isfile(os.path.join(tmpdir, "test_output.csv"))
```
A more meaningful test would be checking the contents of this generated file.

LukasNorbutas replied:
This test is now removed, the contents are checked in the first 2 tests

astanzys-vinted commented on tests/test_scaler.py:
```diff
@@ -0,0 +1,38 @@
+)
+
+
+def test_scaler_output_values(test_input_data, numeric_cols):
```
Let's use mega descriptive names for test cases. It's very helpful when you're not familiar with the domain. For example, I have no idea what standard scaler is supposed to do, and this test name does not give me that info. How about `test_scaler_does_XXXX` and `test_scaler_string_pass` -> `test_scaler_ignores_strings`

LukasNorbutas replied:
Renamed the two remaining tests

astanzys-vinted commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,66 @@
+from typing import List
+
+# Example: standard scaler for numeric input
+def read_csv_to_df(input_data: str, *args, **kwargs) -> pd.DataFrame:
```
Here and everywhere else. Prefixing parameter names with `input` is a bit meaningless. Here you could just use `csv`. `input_df` -> 'df' and so on.

Allardvm replied:
It's also misleading, since the parameter should be the data, but a path. How about just using `path`?

Allardvm commented on .editorconfig:
```diff
@@ -0,0 +1,23 @@
+[*.{py,rst,ini}]
+indent_style = space
+indent_size = 4
+max_line_length = 99
```
Let's stick to 99, and use the `.flake8` file to set black and flake8 to 99 as well.

LukasNorbutas replied:
👍

Allardvm commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+    && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+RUN chmod +x bin/entrypoint.sh
+RUN chmod +x bin/build_image.sh
```
Why would we include `build_image.sh` in the image?

Allardvm commented on component.yaml:
Let's start versioning all of this now so we can get used to that. Do not tie to `:latest`.

LukasNorbutas replied:
We decided to keep the versioning of the component and images for v0.2 of the template. We build YAML files manually for now, we have not yet decided how to handle this in the workflow. I will make new issues about versioning on the template epic for versioning and keep it opened.

Allardvm commented on setup.py:
```diff
@@ -0,0 +1,17 @@
+    name='component_template',
+    description='Template of a kubeflow component',
+    author='LN',
+    install_requires=required,
```
Do we need this? The `requirements.txt` and `install_requires` have different purposes. The former is to build a replicable environment, and thus picks specific versions, the latter specifies a broad range of versions that could work. We need the former use-case—and the latter would be a nice to have, so I'd skip it for now.

Allardvm commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,65 @@
+    logging.getLogger().setLevel(logging.INFO)
+
+    # Parse arguments
+    parser = argparse.ArgumentParser()
```
Would it be worth settling on the use of the `click` package already for CLIs? It's a far better experience as developer, and would be more intuitive for data scientists.

LukasNorbutas replied:
I took a look at click, never used it - seems like a great alternative. I will incorporate it in the next iteration.

Allardvm commented:
Let's add versioning to the components as well. The Python side of it would be to add a `__version__.py` module in the package root containing `__version__ = '0.0.1'`, then importing it in the root `__init__.py` with `from preprocessor.__version__ import __version__`.

The `setup.py` could use something like the below to use it automatically (copied from our internal cookie cutter):
```python
import os

from setuptools import find_packages
from setuptools import setup


def read(filename):
    with open(filename, encoding='utf-8') as f:
        return f.read()


ABOUT = {}
HERE = os.path.abspath(os.path.dirname(__file__))
exec(read(os.path.join(HERE, 'src', '{{cookiecutter.python_package_name}}', '__version__.py')), ABOUT)

setup(
    name='{{cookiecutter.project_name}}',
    version=ABOUT['__version__'],
    description='{{cookiecutter.description}}',
    author='{{cookiecutter.author_name}}',
    packages=find_packages('src'),
    package_dir={'': 'src'},
    long_description=read(os.path.join(HERE, 'README.md')),
    long_description_content_type='text/markdown')
```

LukasNorbutas commented on .dockerignore:
```diff
@@ -0,0 +1,89 @@
+# Byte-compiled / optimized / DLL files
```
Decided to remove dockerignore file, since we are adding specific folders/files in the Dockerfile, not using "add ."

Allardvm replied:
You should still use a dockerignore—it specifies the files to exclude from Docker's internal intermediate layer. Before starting the image build process, any file in the current working directory or lower will be copied into this intermediate layer (the contents of which you'll never see), which slows down the build. Especially if some of these files are large (e.g. local datafiles, models, etc.), I've seen this lead to minute-long delays.

LukasNorbutas commented on .gitignore:
```diff
@@ -0,0 +1,91 @@
+# Byte-compiled / optimized / DLL files
```
👍

LukasNorbutas commented:
Addressed most of the comments here, thanks a lot for a thorough review. A few things left to discuss and include: versioning of the component and the image (+ automated generation of the YAML spec), take a close look at the click lib and use it for parsing arguments. I will create new tasks for these in the epic and address them in v0.2

Allardvm commented on src/preprocessor/standard_scaler.py:
```diff
@@ -0,0 +1,66 @@
+    Read a dataframe and rescale specified columns with standard scaler.
+    """
+    data_copy = input_df.copy()
+    print(cols, type(cols))
```
You'd usually just do the following in every single module:
```python
import logging

LOGGER = logging.getLogger(__name__)

# code
LOGGER.info('foo')
